### PR TITLE
2.2 Assignments Mastin

### PR DESCRIPTION
**Pull request by mmastin**

### CODE REVIEW

**Comment by axrd** [comment]
# Module 01 - Logistic Regression & One-Hot Encoding  

**Overall**: Well done! You're accuracy score is pretty darn good 👍 
**Feedback**: Comment your code! Comment it to force a good habit of discipline. Future you will likely not remember what you meant with every line of code in past notebooks. Keep him in mind. Also, use a validation set, and leverage your coefficients!

## Exploring/Cleaning data
- Try the `df.info()` method. Might help you as an initial step.
- Here's one way you could spot missing values in a functional way:
```
def missing_values(data):
   total = data.isnull().sum()
   percent = data.isnull().sum().sort_values(ascending = False)/len(data)*100
   tp = pd.concat([total, percent], axis=1, keys=['Total', '%'])
   return(np.transpose(tp))
```
- Remember that cleaning data is not just about nulls and NaNs. Read the links in the lecture notebook. 

## Train/Val/Test Split
- Nicely done 👍 

## One-Hot Encoding & Scaling
- Nicely done! Kudos for evaluating cardinality of your features👍 

## Logistic Regression Model
 - Yep. That's how you do it 👍 + 💯 for checking cardinality

## Coefficients
- Great job plotting these. 

## Submit to Kaggle Competition
- Aww yeah! How's it feel to be on the leaderboard?

**Comment by axrd** [comment]
# Module 02 - Decision Trees  

**Overall**: Well done! Congratulations on breaking 80% Accuracy! 📈 
**Feedback**: **_Comment your code!_** Nicely done jumping ahead and trying out a Random 🌲 🌳 🌳 🌲 👍 

## More Exploring/Cleaning data
- Dig deeper. Lookout for missing or incorrect values in other Features. 
- But well done making your wrangling into a function 👍 

## Implement a Decision Tree Classifier
- Yep. Nicely done 👍 + 💯 for going ahead and implementing a forest!
- Plot the feature importances to help guide your approach
```
# Get feature importances
encoder = pipeline.named_steps['onehotencoder']
tree = pipeline.named_steps['decisiontreeclassifier']
feature_names = encoder.transform(X_val).columns
importances = pd.Series(tree.feature_importances_, feature_names)

# Plot feature importances
n = 20
plt.figure(figsize=(10,n/2))
plt.title(f'Top {n} features')
importances.sort_values()[-n:].plot.barh(color='grey');
```

## Submit New Predictions to Kaggle
- I see you climbing the leaderboard 🙌

**Review by brennerhaverlock — CHANGES_REQUESTED** [review]
# Matthew Mastin- 2.2.3 


#This Review has example code

# Part 1. Imports

***
**What went well:**

-Good collective use of imports:

```python
import category_encoders as ce
import numpy as np
import pandas as pd
from sklearn.impute import SimpleImputer
from sklearn.model_selection import train_test_split
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import RobustScaler
from sklearn.tree import DecisionTreeClassifier
pd.set_option('display.float_format', '{:.2f}'.format)


```

**What could be better**

-
-You do not need [] just do df.date_recorded or df.year_recorded or etc: 

```python

def clean(df):
    df = df.copy()
    
    df['date_recorded'] = pd.to_datetime(df['date_recorded'], infer_datetime_format=True)
    
    df['year_recorded'] = df['date_recorded'].dt.year
    df['month_recorded'] = df['date_recorded'].dt.month
    df['day_recorded'] = df['date_recorded'].dt.day
    df = df.drop(columns='date_recorded')
    
    df['years'] = df['year_recorded'] - df['construction_year']
    
    duplicate_columns = ['quantity_group']
    df = df.drop(columns=duplicate_columns)
    
    df['latitude'] = df['latitude'].replace(-2e-08, np.nan)
    
    cols_with_zeros = ['construction_year', 'longitude', 'latitude', 'gps_height',
                      'population']
    for col in cols_with_zeros:
        df[col] = df[col].replace(0, np.nan)
        
    categoricals = df.select_dtypes(exclude='number').columns
    for col in categoricals:
        df[col] = df[col].fillna('MISSING')
        
    return df


```


# What to work on
- [x] Comment more in code 
- [x] Come to PM for questions or help
